govcd/vm_test.go

```diff
@@ -865,6 +869,340 @@ func (vcd *TestVCD) Test_VMPowerOnPowerOff(check *C) {
+		check.Assert(err, IsNil)
+		check.Assert(networkConnectionSection.NetworkConnection[0].Network, Equals, vcd.config.VCD.Networks[0])
+		check.Assert(networkConnectionSection.NetworkConnection[0].IPAddress, Equals, startAddress)
+	}
+
+	if len(vcd.config.VCD.Networks) >= 2 {
+		vapp := vcd.findFirstVapp()
+		vmType, vmName := vcd.findFirstVm(vapp)
+		if vmName == "" {
+			check.Skip("skipping test because no VM is found")
+		}
+
+		fmt.Printf("Running: %s part 2 change network adapter count\n", check.TestName())
+
+		vm := NewVM(&vcd.client.Client)
+		vm.VM = &vmType
+
+		var net OrgVDCNetwork
+		var nets []map[string]interface{}
+		var err error
+
+		// Find networks
+		for index := range vcd.config.VCD.Networks {
+			n := make(map[string]interface{})
+			net, err = vcd.vdc.FindVDCNetwork(vcd.config.VCD.Networks[index])
+
+			check.Assert(err, IsNil)
+			check.Assert(net, NotNil)
+			check.Assert(net.OrgVDCNetwork.Name, Equals, vcd.config.VCD.Networks[index])
+			check.Assert(net.OrgVDCNetwork.HREF, Not(Equals), "")
+
+			n["ip"] = ""
+			n["ip_allocation_mode"] = "POOL"
+			n["is_primary"] = false
+			if index == 0 {
+				n["is_primary"] = true
+			}
+			n["network_name"] = net.OrgVDCNetwork.Name
+
+			// Add missing networks before continue changing them
+			task, err := vapp.AppendNetworkConfig(net.OrgVDCNetwork)
+			check.Assert(err, IsNil)
+			if task != (Task{}) {
+				err = task.WaitTaskCompletion()
+				check.Assert(err, IsNil)
+				task, err = vm.AppendNetworkConnection(net.OrgVDCNetwork)
+				err = task.WaitTaskCompletion()
+				check.Assert(err, IsNil)
+			}
+
+			nets = append(nets, n)
+		}
+
+		// Add networks we found
+		task, err := vm.ChangeNetworkConfig(nets)
+		check.Assert(err, IsNil)
+		err = task.WaitTaskCompletion()
+		check.Assert(err, IsNil)
+		check.Assert(task.Task.Status, Equals, "success")
+
+		// Check if networks were added correctly
+		networkConnectionSection, err := vm.GetNetworkConnectionSection()
+		check.Assert(err, IsNil)
+		for index := range vcd.config.VCD.Networks {
+			check.Assert(networkConnectionSection.NetworkConnection[index].Network, Equals, vcd.config.VCD.Networks[index])
+			check.Assert(networkConnectionSection.NetworkConnection[index].IPAddressAllocationMode, Equals, nets[index]["ip_allocation_mode"])
+			check.Assert(networkConnectionSection.NetworkConnection[index].IPAddress, NotNil)
+		}
+		check.Assert(networkConnectionSection.PrimaryNetworkConnectionIndex, Equals, 0)
+	}
+}
+
+// Test_updateNicParameters_multinic is meant to check functionality of a complicated
+// code structure used in vm.ChangeNetworkConfig which is abstracted into
+// vm.updateNicParameters() method so that it does not contain any API calls, but
+// only adjust the object which is meant to be sent to API. Initially we hit a bug
+// which occurred only when API returned NICs in random order.
+func (vcd *TestVCD) Test_updateNicParameters_multiNIC(check *C) {
```

Comment:
Isn't this function a good candidate for a separate test framework that does not require gocheck? 
It doesn't use `vcd` and could be isolated.

Reply:
Yes it is. And could benefit from subtests. The only thing is to agree how we do assertions. Do we use `if xxxx { t.Fatal` or some assertion lib/framework.

Reply:
Not necessarily `t.Fatal`, could also be `t.Fail`. But the general point. I think it's more of an agreement question rather than a problem.

Reply:
If we plan to produce many such tests, we could create a lightweight assertion library. I have made [one for my own usage in a different project](https://github.com/datacharmer/dbdeployer/blob/master/compare/compare.go) just because I wanted to avoid unnecessary dependencies.

Reply:
So probably the solution now is to keep it light - I will use plain `if` and we evaluate this in future. I don't know if we have many such tests. In my opinion this is good, but can't predict future.

Reply:
OK. I leave it to your judgment. If you can isolate this test without much work, go for it.